### PR TITLE
Product Gallery block: Restrict block to be available only on the Single Product template or the Product Gallery template part

### LINKED ISSUE #11027: Product Gallery: the block should be addable only in the Single Product Template and in the `Product Gallery` template-part
Currently, the Product Gallery is addable on any page. Instead, the block should be addable only on:
- Single Product Template
- Product Gallery Template Part

More details: https://github.com/woocommerce/woocommerce-blocks/pull/10947#issuecomment-1731046705

### PR DESCRIPTION
**Pull request by thealexandrelara**
## What
- Add a TemplateChangeDetector that detects when a template or page changes and notifies any subscribers (currently only the BlockRegistrationManager).
- Add the BlockRegistrationManager, which is responsible for registering/unregistering blocks according to the current template/page.
- In a subsequent PR, we can replace the current implementation of registerBlockSingleProductTemplate with this new pattern, which is more performant and less prone to racing conditions that have already caused issues in the past.

Fixes #11027

## Why
The current implementation of the Product Gallery block does not work outside the Single Product template context (or the Product Gallery template part, which is responsible for the dialog when clicking on the Large Image block). Therefore, we need to restrict the block to be available only within this context.

## Testing Instructions

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

#### Test 1: Single Product template
1. From your WordPress dashboard, go to Appearance > Themes. Select a block-based theme like "Twenty-twenty Three," "Twenty-twenty Four," etc.
2. On the left-hand side menu, click on Appearance > Editor > Templates
3. Find and select the 'Single Product' template from the list.
4. When the Classic Product Template renders, click on Transform into Blocks. This will transform the Classic template in a block template if you haven't done it before.
5. Inside the Page editor, click on the '+' button to add a new block.
6. In the block library that pops up, search for the 'Product Gallery' block. Click on it to add the block to the template.
7. Make sure the Product Gallery block can be added to the editor.
8. Click on Save.
9. Visit a product page of a Variable Product and make sure the block is working as expected.

#### Test 2: Posts
1. From your WordPress dashboard, go to Appearance > Themes. Select a block-based theme like "Twenty-twenty Three," "Twenty-twenty Four," etc.
2. On the left-hand side menu, click on Posts > Add new.
5. Inside the Page editor, click on the '+' button to add a new block.
6. In the block library that pops up, search for the 'Product Gallery' block.
7. Make sure the Product Gallery block DOES NOT appear and cannot be added to the editor.

#### Test 3: Pages
1. From your WordPress dashboard, go to Appearance > Themes. Select a block-based theme like "Twenty-twenty Three," "Twenty-twenty Four," etc.
2. On the left-hand side menu, click on Pages > Add new.
5. Inside the Page editor, click on the '+' button to add a new block.
6. In the block library that pops up, search for the 'Product Gallery' block.
7. Make sure the Product Gallery block DOES NOT appear and cannot be added to the editor.

* [x] Do not include in the Testing Notes 
* [ ] Should be tested by the development team exclusively 

## WooCommerce Visibility

Required:

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental
* [ ] N/A

### CODE REVIEW

File: assets/js/atomic/utils/blocks-registration-manager/block-registration-strategy.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	): boolean;
+	unregister( blockName: string, variationName?: string ): boolean;
+}
+
+export class BlockTypeStrategy implements BlockRegistrationStrategy {
+	register(
+		blockName: string,
+		blockSettings: Partial< BlockConfiguration >
+	): boolean {
+		return Boolean( registerBlockType( blockName, blockSettings ) );
```
Comment (Aljullu) [inline assets/js/atomic/utils/blocks-registration-manager/block-registration-strategy.ts]: There seems to be a TypeScript error in this line. I think the first argument of the function should be a block metadata object instead of the block name.

File: assets/js/atomic/utils/blocks-registration-manager/index.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Internal dependencies
+ */
+import { BlockRegistrationManager } from './blocks-registration-manager';
+import { TemplateChangeDetector } from './template-change-detector';
+
+wp.domReady( () => {
```
Comment (Aljullu) [inline assets/js/atomic/utils/blocks-registration-manager/index.ts]: There is a TS error here too. Maybe this should be [imported from the package](https://github.com/WordPress/gutenberg/tree/b09e9bd32c4bd8740833d7b15ca814ce3c7c1608/packages/dom-ready#api)?

**Review by Aljullu — COMMENTED** [review]
This is testing great! Thanks @thealexandrelara for working on this, it's definitely a tricky topic.

I noticed that with this approach we need to define the blocks with registration restrictions in `blocks-with-restriction.ts`. Do you think it would be possible to create a util wrapping `registerBlockType()` which takes care of that? So we would only need to call that util (ie: `registerBlockTypeWithRestrictions()`) when registering the block instead of having to modify `blocks-with-restriction.ts` if we ever need to do this for other blocks.

By the way, it might be a good idea to add some e2e tests to make sure we don't introduce regressions on this feature in the future. What do you think?

File: assets/js/atomic/utils/blocks-registration-manager/block-registration-strategy.ts (hunk 1 above)
Comment (thealexandrelara) [reply assets/js/atomic/utils/blocks-registration-manager/block-registration-strategy.ts]: Yes, unfortunately the current version that we are using for @woocommerce/blocks is not correctly typed. It accepts both [`blockName` ](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-registration/#block-name)or the [`blockConfiguration`](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-registration/#block-configuration); however, it errors out when using the first one. 

I added the following code to prevent the error from appearing. When we update the package to a version that contains the correct types, an error will occur so we know that we can fix this and remove the comment (this approach is being used in other places in our codebase):
```
// @ts-expect-error: `registerBlockType` is typed in WordPress core
```

File: assets/js/atomic/utils/blocks-registration-manager/index.ts (hunk 2 above)
Comment (thealexandrelara) [reply assets/js/atomic/utils/blocks-registration-manager/index.ts]: Nice catch! I wasn't aware of this package, but I just pushed the changes. Thank you for sharing this with me!

**Comment by thealexandrelara** [comment]
> I noticed that with this approach we need to define the blocks with registration restrictions in `blocks-with-restriction.ts`. Do you think it would be possible to create a util wrapping `registerBlockType()` which takes care of that? So we would only need to call that util (ie: `registerBlockTypeWithRestrictions()`) when registering the block instead of having to modify `blocks-with-restriction.ts` if we ever need to do this for other blocks.

Thank you for reviewing and suggesting this approach! Let me provide more context about why this approach is being used here.

Currently, we have the `assets/js/atomic/utils/register-block-single-product-template.ts`, which works as a wrapper for the `registerBlockType()` and `registerBlockVariation()` functions. One of the issues we encounter when using this wrapper is that, for each component that utilizes it, we subscribe to the data store (core/edit-site and core/edit-post). Consequently, every time the store is updated, this logic is triggered for each subscriber.

We currently have approximately 8 blocks using this approach. Since each instance of registerBlockSingleProductTemplate is isolated from the others, they are triggered independently, but they update the store simultaneously. This can lead to racing conditions (https://github.com/woocommerce/woocommerce-blocks/pull/9090).

The idea behind using this approach is to establish a single source of truth responsible for monitoring and responding to page/post or template changes, as well as registering/unregistering blocks based on the defined restrictions. This approach also prevents racing conditions since there is a single point responsible for block registration/unregistration. Another advantage is improved performance, as we only have one subscriber to the store, which triggers the logic for managing the blocks only when merchants go from one post/page to another or from one template to another.

Regarding your suggestion, what changes do you think we could make to retain these benefits while avoiding the use of a separate file for block restrictions?

> By the way, it might be a good idea to add some e2e tests to make sure we don't introduce regressions on this feature in the future. What do you think?

Sure! I'll be adding some e2e tests for the Product Gallery block to make sure the block is only available inside the Single Product template and not on the posts/pages.

**Comment by roykho** [comment]
I also had similar concerns as Albert. Is there a way to add metadata to the Product Gallery block and when it gets registered through the wrapper, it will pick up on the restrictions instead of using that restriction file?

File: assets/js/atomic/utils/blocks-registration-manager/block-registration-strategy.ts (hunk 1 above)
Comment (Aljullu) [reply assets/js/atomic/utils/blocks-registration-manager/block-registration-strategy.ts]: Thanks for the explanation. It makes sense! :+1:

**Comment by Aljullu** [comment]
> The idea behind using this approach is to establish a single source of truth responsible for monitoring and responding to page/post or template changes, as well as registering/unregistering blocks based on the defined restrictions. This approach also prevents racing conditions since there is a single point responsible for block registration/unregistration. Another advantage is improved performance, as we only have one subscriber to the store, which triggers the logic for managing the blocks only when merchants go from one post/page to another or from one template to another.

That makes a lot of sense, @thealexandrelara! Thanks for the detailed explanation.

I was thinking about something along these lines:

```TS
export const registerBlockTypeWithRestrictions = (
	blockMetadata,
	blockSettings,
	{
		allowedTemplates,
		allowedTemplateParts,
		availableInPostOrPageEditor,
		isVariationBlock,
	}
) => {
	BLOCKS_WITH_RESTRICTION[ blockMetadata.name ] = {
		blockMetadata,
		blockSettings,
		allowedTemplates,
		allowedTemplateParts,
		availableInPostOrPageEditor,
		isVariationBlock,
	};
	registerBlockType( blockMetadata, blockSettings );
};
```

Which then could be called like this:

```TS
registerBlockTypeWithRestrictions(
	productGalleryBlockMetadata,
	ProductGalleryBlockSettings,
	{
		allowedTemplates: {
			'single-product': true,
		},
		allowedTemplateParts: {
			'product-gallery': true,
		},
		availableInPostOrPageEditor: false,
		isVariationBlock: false,
	}
);
```

I didn't test it, and it would probably require some internal refactoring, as `BLOCKS_WITH_RESTRICTION` would be a dynamic constant. So there might be some blockers that I'm not aware of. In any case, I wanted to share it as I think it would make the API much easier to work with.

Besides that, everything looks good to me. :ok_hand:

**Comment by thealexandrelara** [comment]
Thank you for your suggestions, it would be awesome to have this API and it would make much easier to maintain. I was testing your suggestions for improving the API and creating a wrapper around the registerBlockType. However, it appears to be more complex than I initially expected. The issue is that the BlocksRegistrationManager only runs when there's a template or page change. When you first load the page, the function runs as soon as the page loads. However, the blocks are registered later. Since the block isn't registered before the manager runs, the manager doesn't know when it should remove the block, making it available in places where it shouldn't (when loading the page for the first time).

I'll be AFK for the next few weeks, but this PR will be essential for the Product Gallery block. I believe I can improve this API when we start migrating other blocks to use the BlocksRegistrationManager. The other option would be for someone else to continue working on this and improve the API before merging it. What do you think?

P.S.: I've just enabled auto-merge, so once this PR is approved it will be automatically merge.

File: assets/js/atomic/utils/blocks-registration-manager/blocks-registration-manager.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+ */
+export class BlockRegistrationManager
+	implements TemplateChangeDetectorObserver
+{
+	private blocksWithRestriction: BlocksWithRestriction;
+	private unregisteredBlocks: string[] = [];
+	private blockRegistrationStrategy: BlockRegistrationStrategy;
+
+	constructor() {
+		this.blocksWithRestriction = BLOCKS_WITH_RESTRICTION;
```
Comment (Aljullu) [inline assets/js/atomic/utils/blocks-registration-manager/blocks-registration-manager.ts]: Out of curiosity, why assigning `BLOCKS_WITH_RESTRICTION` to a class property instead of using the constant directly?

File: assets/js/atomic/utils/blocks-registration-manager/blocks-registration-manager.ts [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+	 * whether the editor is in post or page mode. It checks whether the current template ID starts with any of the
+	 * allowed templates or template parts for the block, and whether the block is available in the post or page editor.
+	 *
+	 * @param {Object}  params                          - The parameters for the method.
+	 * @param {string}  params.blockWithRestrictionName - The name of the block with restrictions.
+	 * @param {string}  params.currentTemplateId        - The ID of the current template.
+	 * @param {boolean} params.isPostOrPage             - Whether the editor is in a post or page.
+	 * @return {boolean} True if the block should be registered, false otherwise.
+	 */
+	private blockShouldBeRegistered( {
```
Comment (Aljullu) [inline assets/js/atomic/utils/blocks-registration-manager/blocks-registration-manager.ts]: Nit: I think it would make it more clear to rename this method as `shouldBlockBeRegistered`. It returns a boolean so it's like a question, right?

File: assets/js/atomic/utils/blocks-registration-manager/blocks-registration-manager.ts [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+				this.blockRegistrationStrategy = this.blocksWithRestriction[
+					blockWithRestrictionName
+				].isVariationBlock
+					? new BlockVariationStrategy()
+					: new BlockTypeStrategy();
+
+				this.blockRegistrationStrategy.unregister(
+					blockWithRestrictionName
+				);
+				unregisterBlockType( blockWithRestrictionName );
```
Comment (Aljullu) [inline assets/js/atomic/utils/blocks-registration-manager/blocks-registration-manager.ts]: I'm a bit confused by this, but ultimately these two instructions do the same, no? We are trying to unregister the block twice, causing a JS error:

imatge

**Review by Aljullu — COMMENTED** [review]
Makes sense, thanks @thealexandrelara! Given that it's not a public API, I think it's fine to iterate on it after merge. I left some last comments here and there, but everything else LGTM.

I will go ahead and push some fixes directly to this PR, as I think at least the JS error is a blocking one.

File: assets/js/atomic/utils/blocks-registration-manager/blocks-registration-manager.ts (hunk 3 above)
Comment (Aljullu) [reply assets/js/atomic/utils/blocks-registration-manager/blocks-registration-manager.ts]: I removed this property in 904f9464becc9004ccaedc0e4d6838fed97e4815.

File: assets/js/atomic/utils/blocks-registration-manager/blocks-registration-manager.ts (hunk 4 above)
Comment (Aljullu) [reply assets/js/atomic/utils/blocks-registration-manager/blocks-registration-manager.ts]: Method renamed in 08eafcb2020ad5738beeb11604f20072fc2dc172.

File: assets/js/atomic/utils/blocks-registration-manager/blocks-registration-manager.ts (hunk 5 above)
Comment (Aljullu) [reply assets/js/atomic/utils/blocks-registration-manager/blocks-registration-manager.ts]: This should be fixed in 6c809cf92ed85812eab292655b3699568054cafe.

**Comment by Aljullu** [comment]
@roykho I pushed some additional commits to this PR. Would you mind taking a second look to make sure I didn't mess anything up? :pray:

**Comment by roykho** [comment]
@Aljullu - it's working correctly thanks for the additions.

**Review by roykho — APPROVED** [review]
Thanks for working on this. Everything is testing as described even the e2e tests. Great job!
